### ISSUE TITLE
Strict comments option for indent rule

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#9017]
**What rule do you want to change?**

`indent`

**Does this change cause the rule to produce more or fewer warnings?**

Potentially more warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option:

```json
{
    "indent": ["error", 4, { "comments": "strict" }]
}
```

This new option will enforce that lines starting with comments should be indented where a token on the same line would normally be expected (i.e., removes the tolerance for comments aligned with first token/comment of line above or below).

**Please provide some example code that this change will affect:**

```js
doSomething([
// Comment should be indented, but currently is not
    "foo",
    "bar",
// Comment should be indented, and currently is
    "baz"
// Comment should be indented, but currently is not
]);
```

**What does the rule currently do for this code?**

First and third comments are not reported or fixed. Second comment is reported and fixed.

**What will the rule do after it's changed?**

All three comments are reported and fixed. See below.

```js
doSomething([
    // Comment should be indented, but currently is not
    "foo",
    "bar",
    // Comment should be indented, and currently is
    "baz"
    // Comment should be indented, but currently is not
]);
```

**Comment by platinumazure** [comment]
And I'll champion this.

**Comment by gyandeeps** [comment]
I dont like the idea of having strict or loose type terminology in indent. I think it will lead to more confusion over time. A think in indent is either indented correctly or not should be the rule.

Can you please share the expected fixed code for clarity.

**Comment by platinumazure** [comment]
@gyandeeps I'm certainly open to bikeshedding on the name, but the fact is, right now our comment indentation can only be described as "loose" because it allows 3 indentations:

* Match indentation of line above;
* Match indentation of line below;
* "Expected" location (i.e., what would be enforced if this were a token rather than a comment).

The goal of this proposal is to allow users to opt into comment indentation rules that match the indentation rules already used for tokens in the same place. I see that as "strict" compared to the current rules. But again, I am open to alternative names.

**Comment by not-an-aardvark** [comment]
1. Do you anticipate that users would be okay with this indentation? That's the main reasoning for the current behavior.

    ```js
    if (foo) {
        // comment describing bar()
    } else if (bar()) {
    }
    ```

1. It should be noted that the rule currently always ignores comments like this:

    ```js
    switch (foo) {
        case bar:
            baz;
        // no default
    }
    ```

    The `// no default` comment doesn't match any of the "allowed" intentations for it. (The token before has 8 spaces of indentation, the token after has 0 spaces, and the "default" indentation would be 8 spaces.) To get around this issue with the current behavior, the indentation of comments after the last switch case is [always completely ignored](https://github.com/eslint/eslint/blob/9b6c5529c797395ff880e60d2bfaada21024e1e9/lib/rules/indent.js#L1257-L1261).

1. I'm also not a big fan of the `strict` option name (it seems like in the future we might have a different option that also makes the rule more strict with regard to comments). That said, I don't have any better alternatives off the top of my head.

**Comment by platinumazure** [comment]
@not-an-aardvark 

> Do you anticipate that users would be okay with this indentation? That's the main reasoning for the current behavior.

Yes-- my assumption is that anyone who wants to use this style would use either a trailing comment, or a comment immediately inside the block at the right indentation. That said, this being a new option, users can also simply choose not to use it.

```js
if (foo) {
} else if (bar()) {   // trailing comment describing bar()
    // or, comment describing bar() or implications of bar()
}
```

> It should be noted that the rule currently always ignores comments [at the end of switch statements]:

I would be okay with continuing to ignore those, if needed. (A better solution, IMO, would be to allow a comment to match the case's alignment if there is no code following the comment. Or, we can encourage users to consider using blocks around their case statement bodies, which would allow `// no default` to match with a closing brace.

**Comment by platinumazure** [comment]
Closing due to lack of interest in implementing.
